### ISSUE TITLE
Does stacking regressor work with pipeline with hyper parameter search?

### ISSUE THREAD

**Issue by ritviksahajpal** [rasbt/mlxtend#313]
Thanks for a great package! I had a query about whether stacking regressor supports a pipeline where I use a algorithm for feature selection. Sample code:

```
        from sklearn.svm import SVR
        from sklearn.ensemble.forest import RandomForestRegressor as RFR
        from mlxtend.regressor import StackingCVRegressor
        from pyearth import Earth
        from skgarden import RandomForestQuantileRegressor as QFR

        svr_rbf = SVR(kernel='rbf')  # meta-regressor
        qfr = QFR(n_estimators=1250, n_jobs=cc.ncpu, random_state=0)
        earth = Earth(max_degree=5, minspan_alpha=.05, endspan_alpha=.05, max_terms=10, check_every=1, thresh=0.)

        regressors = [earth, qfr]
        self.model = StackingCVRegressor(regressors=regressors, meta_regressor=svr_rbf)
        self.param_grid.update({'clf__qfr__max_features': ['auto', 'log2'],
                                'clf__qfr__n_estimators': [550, 950],
                                'clf__qfr__min_samples_split': [2, 3, 6, 9],
                                'clf__qfr__max_depth': [3, 6, 9, 12],
                                'clf__qfr__min_samples_leaf': [1, 3, 6, 9, 12, 15, 18],
                                'clf__meta-svr__C': [0.1, 1.0, 10.0, 100.0],
                                'clf__meta-svr__gamma': [0.1, 1.0, 10.0]})
```

The above code sets up self.model which is used as below:

```
         from sklearn.pipeline import Pipeline
         pipeline = Pipeline([('fs', feat_selection), ('clf', self.model), ])
         gs = BayesSearchCV(pipeline, self.param_grid, verbose=1, error_score=np.nan)
```

However, this leads to this error:
```

    ValueError: Invalid parameter qfr for estimator StackingCVRegressor(cv=5,
              meta_regressor=SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='auto',
      kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False),
              regressors=[Earth(allow_linear=None, allow_missing=False, check_every=1,
       enable_pruning=True, endspan=None, endspan_alpha=0.05, fast_K=None,
       fast_h=None, feature_importance_type=None, max_degree=5, max_terms=10,
       min_search_points=None, minspan=None, minspan_alpha=0.05, penalty=None,
       smoo...jobs=7,
                   oob_score=False, random_state=0, verbose=0,
                   warm_start=False)],
              shuffle=True, store_train_meta_features=False,
              use_features_in_secondary=False). Check the list of available parameters with `estimator.get_params().keys()`.
```

Is there something wrong in the approach I follow? The error goes away when I remove all the `clf__qfr*` elements from `self.param_grid`. Is there a way to retain them since I would like to do a hyper parameter search

**Comment by rasbt** [comment]
Hey there! At a quick glance, I can't figure out where your "clf__" prefix comes from, maybe that's the culprit? I.e., maybe just try

```python
{'qfr__max_features': ['auto', 'log2'],
 'qfr__n_estimators': [550, 950], 
 ...
```

Cheers,
Sebastian

**Comment by rasbt** [comment]
Hey there!
It should support pipelines. Looking at your example, I can't figure out why it wouldn't work. E.g., see the following self-contained example, which works fine

```python
from mlxtend.regressor import StackingCVRegressor
from sklearn.datasets import load_boston
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import make_pipeline
from sklearn.pipeline import Pipeline

X, y = load_boston(return_X_y=True)

ridge = Ridge()
lasso = Lasso()
rf = RandomForestRegressor()

stack = StackingCVRegressor(regressors=(lasso, ridge),
                            meta_regressor=rf, 
                            use_features_in_secondary=True)

pipeline = make_pipeline(StandardScaler(), stack)

params = {'stackingcvregressor__lasso__alpha': [0.1, 1.0, 10.0],
          'stackingcvregressor__ridge__alpha': [0.1, 1.0, 10.0]}

grid = GridSearchCV(
    verbose=1,
    estimator=pipeline, 
    param_grid=params, 
    cv=5,
    refit=True
)

grid.fit(X, y)

print("Best: %f using %s" % (grid.best_score_, grid.best_params_))
```

Instead of using `make_pipeline`, you could also do

```python
pipeline = Pipeline([('scaler', StandardScaler()), ('model', stack)])

params = {'model__lasso__alpha': [0.1, 1.0, 10.0],
          'model__ridge__alpha': [0.1, 1.0, 10.0]}
```

Not sure, but maybe for debugging, could you try scikit-learn's `GridSearchCV` and see if that would work? (And maybe also try `make_pipeline`). In any case, once we figured out what's going on, it would be good to add a pipeline example to the docs :)

**Comment by ritviksahajpal** [comment]
thanks @rasbt , I checked with `GridSearchCV` but get the same error. I think it might be due to the fact that I have an additional step in the pipeline to select features. I will keep you posted here

**Comment by ritviksahajpal** [comment]
@rasbt I think I found the issue with my code. so instead of using this in the param_grid:

`clf__qfr__max_features`, I should have used `'clf__randomforestquantileregressor__max_features`. This is because of how I was importing qfr:

`from skgarden import RandomForestQuantileRegressor as QFR`. 

Thank you so much again for your help and also an excellent library!

**Comment by rasbt** [comment]
Oh, yeah, that makes sense. Somehow haven't thought about it. It's the way the "naming" was implemented such that it used the name of the class instead of the name that is assigned inside the pipeline. When we implemented it, there were plans in scikit-learn to adopt this way, too, but yeah, it never got to that.

Anyways, I think I can close this issue then. But let me know if you have any other questions :)
